GVFS/GVFS.Common/NamedPipes/QueuedPipeWriter.cs

```diff
@@ -0,0 +1,209 @@
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.IO.Pipes;
+using System.Threading;
+
+namespace GVFS.Common.NamedPipes
+{
+    public class QueuedPipeWriter : IDisposable
+    {
+        private static readonly TimeSpan PollingInterval = TimeSpan.FromSeconds(30);
+
+        private readonly Func<NamedPipeClientStream> pipeFactory;
+        private readonly ConcurrentQueue<byte[]> queue = new ConcurrentQueue<byte[]>();
+        private readonly AutoResetEvent queuePulse = new AutoResetEvent(initialState: false);
+
+        private Thread writerThread;
+        private NamedPipeClientStream pipeClient;
+        private bool isRunning;
+        private bool isDisposed;
+
+        public QueuedPipeWriter(Func<NamedPipeClientStream> pipeFactory)
+        {
+            this.pipeFactory = pipeFactory;
+        }
+
+        public event EventHandler<QueuedPipeWriterStateEventArgs> StateChanged;
+
+        public enum QueuedPipeWriterState
+        {
+            Stopped,
+            Failing,
+            Healthy,
+        }
+
+        public void Start()
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeWriter));
+            }
+
+            if (this.isRunning)
+            {
+                return;
+            }
+
+            this.isRunning = true;
+
+            this.writerThread = new Thread(this.ThreadProc)
+            {
+                Name = nameof(QueuedPipeWriter),
+                IsBackground = true,
+            };
+
+            this.writerThread.Start();
+        }
+
+        public void Enqueue(byte[] data)
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeWriter));
+            }
+
+            if (!this.isRunning)
+            {
+                throw new InvalidOperationException($"{nameof(QueuedPipeWriter)} is not started");
+            }
+
+            this.queue.Enqueue(data);
```

Comment:
No locking is required here as this is a `ConcurrentQueue<>` 😃

Reply:
Also note that I've set _no upper limit to the size of this queue_! This means that if many hundreds of threads send messages we'll have a lot of messages to churn through. Should I set an upper limit? Opinions for a magic number if we want this?

Note: we always dequeue messages if there are any, and if we fail to deliver them over the pipe we never requeue it.. in other words the queue will never grow forever.. it is always attempting to be drained.
(Plus if the `gvfs.telemetry-pipe` configuration variable isn't set we're never even executed to start with!)

Reply:
I think we should test this and see what the memory pressure is at how many messages.  If the dequeuing thread dies for some reason we would be queuing and it would grow forever.  Raises the question whether we should have a `try...catch` in the `ThreadProc`?

Reply:
There is now an upper limit to the queue of 256 messages. Given in testing, messages are almost immediately dispatched (and each message is typically only about 400-800 _bytes_ in size), I don't expect there to be a backlog of unsent (or not-yet-dropped) messages.